es6-shim.js

```diff
@@ -289,11 +289,16 @@
       if (x === void 0 || x === null || x === true || x === false) {
         return false;
       }
-      var type = typeof x;
-      if (type === 'string' || type === 'number' || type === 'symbol') {
-        return false;
-      }
-      return type === 'function' || type === 'object';
+      // This might miss the "(non-standard exotic and does not implement
+      // [[Call]])" case from
+      // http://www.ecma-international.org/ecma-262/6.0/#sec-typeof-operator-runtime-semantics-evaluation
+      // but we can't find any evidence these objects exist in practice.
+      // If we find some in the future, you could test `Object(x) === x`,
+      // which is reliable according to
+      // http://www.ecma-international.org/ecma-262/6.0/#sec-toobject
+      // but is not well optimized by runtimes and creates an object
+      // whenever it returns false, and thus is very slow.
+      return typeof x === 'function' || typeof x === 'object';
```

Comment:
That's an interesting proposal.  You'd have to add back the string/number/symbol checks in that case, though, to ensure that only "exotics" hit the `|| Object(x)` case.  Not sure the extra complexity is worth it, given that I can't actually identify any valid "exotics" in any implementation that don't return 'object' or 'function'.  Presumably if there really is some instance in the wild someone will file a bug complaining that es6-shim throws a `TypeError` in some case when it shouldn't, and we can tweak the check then.